test/torch_optimizer_test.py

```diff
@@ -0,0 +1,525 @@
+# Copyright 2020 Bluefog Team. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+# ==============================================================================
+
+# Note this end-to-end only covers the BlueFog optimizers for a single machine,
+# all the hierarchical cases are not fully under test yet.
+
+from __future__ import absolute_import
+from __future__ import division
+from __future__ import print_function
+
+import pytest
+import numpy as np
+import torch
+import torch.nn as nn
+import torch.optim as optim
+from torch.utils.data import Dataset, DataLoader
+
+import bluefog.torch as bf
+from bluefog.common import topology_util
+
+TEST_ON_GPU = torch.cuda.is_available()
+
+# A linear model for testing
+class LinearNet(nn.Module):
+    def __init__(self, input_dim, output_dim):
+        super(LinearNet, self).__init__()
+        self.fc = nn.Linear(input_dim, output_dim)
+
+    def forward(self, x):
+        return self.fc(x)
+
+# A Simple dataset for testing.
+class SimpleDataset:
+    def __init__(self, x, y):
+        self._x = x
+        self._y = y
+    
+    def __len__(self):
+        return len(self._y)
+    
+    def __getitem__(self, idx):
+        return (torch.tensor(self._x[idx],dtype=torch.float32),
+                torch.tensor(self._y[idx],dtype=torch.float32))
+
+# A ProblemBuilder for the linear problem with a specified input and output dimension.
+# The matrix A are randomly generated now.
+#   y = AX + e, e ~ N(0, noise_level^2)
+class LinearProblemBuilder:
+    def __init__(self, input_dim = 16, output_dim = 3, noise_level = 1e-5):
+        self._input_dim = input_dim
+        self._output_dim = output_dim
+        self._noise_level = noise_level
+        self._matrix_gen_seed = 0
+        self._generate_matrices()
+
+    def _generate_matrices(self):
+        state = np.random.get_state()
+        np.random.seed(self._matrix_gen_seed)
+        self._A = np.random.randn(self._input_dim, self._output_dim)
+        #self._A = np.ones((self._input_dim, self._output_dim))
+        np.random.set_state(state)
+
+    @property
```

Comment:
In my opinion, the setter and getter pattern in the test utility function is not necessary. But it is good practice

Reply:
Done